### ISSUE TITLE
Support for HIP/ROCm

### ISSUE THREAD

**Issue by peastman** [conda-forge/conda-forge.github.io#1923]
### Your question:

I'm interested in distributing packages that use HIP, AMD's framework for GPU compute.  It's basically their equivalent of CUDA.  There are a couple of packages on conda-forge related to it (https://anaconda.org/conda-forge/hip, https://anaconda.org/conda-forge/rocm-comgr), but neither one has been updated in several years.

Is there a currently supported way to build packages that use HIP?

If not, are there any plans to add support in the future?

**Comment by isuruf** [comment]
No plans. PRs are welcome to update those packages.

**Comment by isuruf** [comment]
> Is this way appropriate to be used with the conda-forge infrastructure?

No, we prefer to build all the dependencies as conda packages.

**Comment by peastman** [comment]
Would it be possible to have ROCm components preinstalled in the image, the same way that's been done for CUDA?

**Comment by isuruf** [comment]
CUDA was done that way because it is proprietary software and we couldn't re-distribute it. However, we are distributing the CUDA compilers since CUDA 12 with permission from NVIDIA and once we drop CUDA 11, we will not preinstall CUDA at all.

ROCM is not proprietary and we already have really old version of ROCM packaged. So, it's not impossible. We definitely don't want to preinstall them when it's possible to package ROCM.

**Comment by zklaus** [comment]
There is renewed effort on getting ROCm into conda-forge. The involved feedstocks are shown below. Currently we are updating to the latest ROCm 6.1.0 and will move on to some basic libraries hopefully soon.

```mermaid
%%{init: {'securityLevel': 'loose'}}%%
flowchart TD
  subgraph hip-feedstock
  hip-runtime-amd
  hip-devel --> hip-runtime-amd
  hipcc --> hip-devel
  rocm-opencl
  rocm-opencl-runtime --> rocm-opencl
  end
  hip-feedstock --> rocminfo & rocr-runtime & rocm-comgr & rocm-cmake
  rocm-cmake
  rocminfo --> rocr-runtime
  rocm-comgr --> rocm-device-libs & rocm-cmake
  rocr-runtime --> roct-thunk-interface & rocm-device-libs
  rocm-device-libs --> rocm-cmake
  roct-thunk-interface
  
  click hip-runtime-amd "https://github.com/conda-forge/hip-feedstock" "hip-feedstock"
  click hip-devel "https://github.com/conda-forge/hip-feedstock" "hip-feedstock"
  click hipcc "https://github.com/conda-forge/hip-feedstock" "hip-feedstock"
  click rocm-opencl "https://github.com/conda-forge/hip-feedstock" "hip-feedstock"
  click rocm-opencl-runtime "https://github.com/conda-forge/hip-feedstock" "hip-feedstock"
  click rocminfo "https://github.com/conda-forge/rocminfo-feedstock" "rocminfo-feedstock"
  click rocr-runtime "https://github.com/conda-forge/rocr-runtime-feedstock" "rocr-runtime-feedstock"
  click rocm-comgr "https://github.com/conda-forge/rocm-comgr-feedstock" "rocm-comgr-feedstock"
  click roct-thunk-interface "https://github.com/conda-forge/roct-thunk-interface-feedstock" "roct-thunk-interface-feedstock"
  click rocm-device-libs "https://github.com/conda-forge/rocm-device-libs-feedstock" "rocm-device-libs-feedstock"
  click rocm-cmake "https://github.com/conda-forge/rocm-cmake-feedstock" "rocm-cmake-feedstock"
```

**Comment by devtomschi** [comment]
@zklaus, @isuruf How are things progressing towards 6.1.0? I'm also interested in an updated HIP 6.1+, so let me know if there is something I can help you with.

**Comment by zklaus** [comment]
@traversaro, that would be very welcome!

More broadly, I have been thinking to try and align conda-forge's rocm packages better with [the upstream guidance](https://rocm.docs.amd.com/projects/install-on-linux/en/latest/reference/package-manager-integration.html).

The tricky bit is, imho, to figure out how to distribute this over feedstocks to minimize hassle.

We should probably also take [the upcoming build system](https://github.com/ROCm/TheRock) into account.

**Comment by traversaro** [comment]
> More broadly, I have been thinking to try and align conda-forge's rocm packages better with [the upstream guidance](https://rocm.docs.amd.com/projects/install-on-linux/en/latest/reference/package-manager-integration.html).

Thanks for the reference, I was not aware of it.

> The tricky bit is, imho, to figure out how to distribute this over feedstocks to minimize hassle.

I can look into it and try to formulate a proposal.

> We should probably also take [the upcoming build system](https://github.com/ROCm/TheRock) into account.

I had the impression (but I may be wrong) that `TheRock` was a classical cmake-based superbuild/virtual monorepo system, that is mostly useful for people working on rocm itself to easily compile from source and using the development version of all the dependencies, but not something that is useful to use directly in downstream package managers, that all have their own ways of build many packages. To do a parallelism, it is a bit like https://gitlab.kitware.com/vtk/vtk-superbuild, that is useful if you work on vtk, but we do not used it in any way in https://github.com/conda-forge/vtk-feedstock .

**Comment by traversaro** [comment]
By the way, given that rocm 7 should be out in ~September/October 2025 (see https://www.amd.com/en/blogs/2025/enabling-the-future-of-ai-introducing-amd-rocm-7-and-the-amd-developer-cloud.html), perhaps it could make sense to focus directly on packaging rocm 7 ?

**Comment by traversaro** [comment]
> I would like to (eventually) test this on a `gfx90a` based system. [@traversaro](https://github.com/traversaro) do you expect to support such "old" AMD Instinct GPUs as well?

Yes, I think a good default list for targets to support for may be:

~~~
gfx908;gfx90a;gfx942;gfx950;gfx1100;gfx1101;gfx1103;gfx1150;gfx1151;gfx1200;gfx1201
~~~

To obtain this, I basically took the default list of hipblaslt (see https://github.com/ROCm/rocm-libraries/blob/rocm-7.0.1/projects/hipblaslt/CMakeLists.txt#L164), remove `xnack` specifiers (those can be added on specific libraries if necessary for performance reasons, and removed `gfx1103` as `rocblas` does not support it (see https://github.com/traversaro/rock-the-conda/issues/3#issuecomment-3317702879). This basically indeed boils down to the targets officially supported (see https://rocm.docs.amd.com/en/latest/compatibility/compatibility-matrix.html) with the additional targets `1150` and `1151` (Strix Point and Strix Halo, that I am personally interested in.

`gfx908` and `gfx90a` were temporary removed in this POC packages as they required a patch tricky to apply for rocBLAS to build on on them, see https://github.com/traversaro/rock-the-conda/issues/3#issuecomment-3317381230 . The Gentoo project already have a working patch https://github.com/gentoo/gentoo/pull/42887, so we can apply that to support also `gfx908` and `gfx90a`. I just avoided that to quickly reach a point in which I could test an actual workload to validate the use of stock clang 20.

**Comment by traversaro** [comment]
> Thanks for all the work, [@traversaro](https://github.com/traversaro)! Do I understand you correctly that this is all about version 7? In other words, do you agree that it would be good to lift all of the existing feedstocks to the latest version 6?

Yes, I initially looked into packaging 6 but I was blocked by https://github.com/conda-forge/rocr-runtime-feedstock/pull/23#issuecomment-3292580469, that was not something trivial to solve. Packaging ROCm 7 was much easier process, so I think it make sense to just update all the feedstocks to ROCm 7 and built on top of that.

**Comment by traversaro** [comment]
PRs for updates to ROCm 7 (I plant to update this issue to have a single place were to track all the PRs):

* [x] `rocm-core` : https://github.com/conda-forge/rocm-core-feedstock/pull/4 
* [x] `rocm-cmake` : https://github.com/conda-forge/rocm-cmake-feedstock/pull/25 
* [x] `rocm-devices-libs` : https://github.com/conda-forge/rocm-device-libs-feedstock/pull/22
* [x] `rocm-comgr`
* [x] `rocr-runtime`
* [ ] `rocminfo`
* [ ] `hip`
* [x] `rocm-smi` : https://github.com/conda-forge/rocm-smi-feedstock/pull/20
* [ ] `rocprim`
* [ ] `rocfft`
* [ ] `hipfft`
* [ ] `roctracer`
* [ ] `hipblas-common`
* [ ] `hiblaslt`
* [ ] `rocblas`
* [ ] `rocsolver`
* [ ] `hipblas`

**Comment by traversaro** [comment]
> Are there instructions on how to build packages that depend on ROCm? I've been trying to figure it out, but so far without success. In the `host` section of my recipe I include `rocm-core`, `rocm-cmake`, `hipcc`, and `hip-devel`. When running cmake I include the option `-DCMAKE_MODULE_PATH=${PREFIX}/lib/cmake/hip`. I can tell it's successfully finding the cmake files, because I get a deprecation warning coming from `hiprtc-config.cmake`. But cmake fails with a message that the target `hip::host` can't be found.

See https://github.com/conda-forge/hip-feedstock/issues/21#issuecomment-3337150276 . I plan to improve this in the 7.0.0 migration.

**Comment by peastman** [comment]
Thanks!  Adding the other packages fixed it.

**Comment by traversaro** [comment]
Another point that is interesting to discuss is how we should version rocm libraries like rocfft, hipfft, rocblas etc etc, see https://github.com/conda-forge/rocfft-feedstock/issues/1 .

**Comment by traversaro** [comment]
Ok, there was a bit of update on some issues. W.r.t. to what was reported in https://github.com/conda-forge/conda-forge.github.io/issues/1923#issuecomment-3324588344, there was some more work.

In particular:
* the recipes have been updated to package rocm 7.0.2
* the architecture list has been expanded to also include `gfx908` and `gfx90a` on top of all architectures listed in https://github.com/conda-forge/conda-forge.github.io/issues/1923#issuecomment-3324588344, and the list is centralized in https://github.com/traversaro/rock-the-conda/blob/main/recipes/hip/activate/hip-clang_activate.sh#L17C53-L17C59, to avoid redefining it for each recipe
* the hip dependencies have been cleaned up to use the compiler conda-forge infrastructure, so now all the HIP-using feedstocks just have a `{{ compiler('hip') }}` or `${{ compiler('hip') }}` build dependency if they need to compile HIP code, see https://github.com/traversaro/rock-the-conda/blob/rev2/recipes/hipfft/recipe.yaml for an example of a typical HIP library recipe.

If you want to test this yourself (also on `gfx908` and `gfx90a` architectures), just run:

~~~
git clone https://github.com/traversaro/rock-the-conda
cd rock-the-conda/examples/llama.cpp
git checkout rev2
pixi run benchmark
~~~

Upstreaming this in conda-forge is tracked in https://github.com/conda-forge/conda-forge.github.io/issues/1923#issuecomment-3339551001, and the next PR in the sequence is https://github.com/conda-forge/rocm-cmake-feedstock/pull/25 .

**Comment by traversaro** [comment]
Good news everyone! 

Together with @flferretti we iterated a lot on https://github.com/gbionics/rock-the-conda, and we reached a stage in which we have packages for llama.cpp and pytorch built for rocm that are ready to be tested (we tested them on gfx942, gfx1150 and gfx1151).

To quickly check if the package work fine on your system with an AMD gpu, just run:

~~~
git clone https://github.com/gbionics/rock-the-conda
cd rock-the-conda/examples/llama.cpp
git checkout rev3
pixi run benchmark
~~~

or

~~~
git clone https://github.com/gbionics/rock-the-conda
cd rock-the-conda/examples/pytorch
git checkout rev3
pixi run benchmark
~~~

and verify that your GPU is being used.

To install these packages with conda, just run:

~~~
conda create -n testrocmllama -c https://prefix.dev/rock-the-conda -c conda-forge llama-cpp=*=*rocm*
~~~

or 

~~~
conda create -n testrocmpytorch -c https://prefix.dev/rock-the-conda -c conda-forge pytorch=*=*rocm*
~~~

For more info or report problems, please check https://github.com/gbionics/rock-the-conda .

Obviously the long term plan is to upstream all of this to conda-forge. The next PR in the series of update PRs is https://github.com/conda-forge/rocm-device-libs-feedstock/pull/25 , and prelimary comments are welcome to any recipe in https://github.com/gbionics/rock-the-conda, in particular to the `hip-feedstock` recipe in https://github.com/gbionics/rock-the-conda/blob/main/recipes/hip/meta.yaml that add support for `${{ compiler('hip') }}`, similar to how other programming languages are dealt with on conda-forge.

**Comment by zklaus** [comment]
Thank you so much, @traversaro 🎉!

I had a look at the hip recipe and it is a clear improvement over what is on the feedstock today, in particular the `hip-clang_linux-64` compiler package, the fixed `hipcc` run dependencies, and the real compile tests.

With conda-forge/rocm-comgr-feedstock#21 merged, every host dependency of `hip` is now at 7.2.4 on conda-forge: rocm-core, rocm-cmake, rocm-device-libs, rocr-runtime, rocminfo, rocm-smi and rocm-comgr. As far as I can tell nothing blocks `hip` anymore.

Could you open the recipe as a PR against conda-forge/hip-feedstock at 7.2.4?

Once the hip PR is in, the next steps would be the pinning PR and then rocfft/hipfft at the ROCm version, per conda-forge/rocfft-feedstock#1.
